### PR TITLE
RPC: starknet_simulateTransactions endpoint

### PR DESCRIPTION
**Pull request by m-kus**
The goal is to enable `starknet_simulateTransactions` endpoint from the new [trace spec](https://github.com/starkware-libs/starknet-specs/blob/master/api/starknet_trace_api_openrpc.json#L41) that allows to dry run a batch of transactions (declare, deploy, invoke) and get execution trace (with all the internal calls).

Depends on:
- https://github.com/keep-starknet-strange/starknet-api/pull/10
- https://github.com/keep-starknet-strange/cairo-rs/pull/3
- https://github.com/keep-starknet-strange/blockifier/pull/9
- https://github.com/0xs34n/starknet.js/pull/694
- https://github.com/keep-starknet-strange/madara/pull/992

Notes:
- Supports both `starknet_simulateTransaction` (legacy starknet.js compatible) and `starknet_simulateTransactions` (spec 0.4.0-rc2)
- Several RPC types can be replaced once `starknet-rs` supports spec `0.4.0` 
   https://github.com/xJonathanLEI/starknet-rs/issues/437

Big thanks to @tdelabro and @greged93 for the help :heart: 

Cc @Eikix 

Created in Starknet Hacker House :croissant:

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -134,10 +134,11 @@ starknet-crypto = { version = "0.6.0", default-features = false }
 starknet-core = { version = "0.5.0", default-features = false }
 starknet-ff = { version = "0.3.4", default-features = false }
 
-blockifier = { git = "https://github.com/keep-starknet-strange/blockifier", branch = "no_std-support", default-features = false, features = [
+# TODO: switch me back to keep-starknet-strange org
+blockifier = { git = "https://github.com/m-kus/blockifier", branch = "no_std-support", default-features = false, features = [
```
Comment (LucasLvy) [inline Cargo.toml:138]: why use your fork ?

File: Cargo.toml [hunk 2]
```diff
@@ -134,10 +134,11 @@ starknet-crypto = { version = "0.6.0", default-features = false }
...
   "parity-scale-codec",
 ] }
-starknet_api = { git = "https://github.com/keep-starknet-strange/starknet-api", branch = "no_std-support", features = [
+starknet_api = { git = "https://github.com/m-kus/starknet-api", branch = "no_std-support", features = [
```
Comment (LucasLvy) [inline Cargo.toml:141]: why use your fork ?

File: crates/client/rpc-core/src/lib.rs [hunk 3]
```diff
@@ -145,4 +146,25 @@ pub trait StarknetRpcApi {
     /// contract.
     #[method(name = "getProof")]
     fn get_proof(&self, get_proof_input: RpcGetProofInput) -> RpcResult<RpcGetProofOutput>;
+
+    /// Simulate a given sequence of transactions on the requested state, and generate the execution
+    /// traces. If one of the transactions is reverted, raises error.
+    /// Spec version 0.4.0-rc2
```
Comment (LucasLvy) [inline crates/client/rpc-core/src/lib.rs:152]: don't specify that plz

File: crates/client/rpc-core/src/lib.rs [hunk 4]
```diff
@@ -145,4 +146,25 @@ pub trait StarknetRpcApi {
...
+    fn simulate_transactions(
+        &self,
+        block_id: BlockId,
+        transactions: Vec<BroadcastedTransaction>,
+        simulation_flags: Vec<SimulateTransactionFlag>,
+    ) -> RpcResult<Vec<SimulateTransactionResult>>;
+
+    /// Compatibility with Starknet.js
+    /// See https://github.com/0xs34n/starknet.js/blob/829f5f4375a63c7c269e34dad22c9cd8874a64ba/src/provider/rpc.ts#L546
+    #[method(name = "simulateTransaction")]
```
Comment (LucasLvy) [inline crates/client/rpc-core/src/lib.rs:163]: wdym ? this endpoint isn't in the specs ?

File: crates/client/rpc-core/src/trace.rs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        // TODO: actual_resources are not passed atm because it's tricky to implement SCALE on hashmap
+        // Alternatively we can aggregate consumed gas from all the call_info structures
+        gas_consumed: 0,
+    }
+}
+
+/// Converts StarkFelt used in Blockifier/StarknetAPI to FieldElement from starknet-rs
+pub fn convert_felt(felt: StarkFelt) -> FieldElement {
+    FieldElement::from_byte_slice_be(felt.bytes()).expect("Expected just to be diff representations of [u8; 32]")
+}
```
Comment (LucasLvy) [inline crates/client/rpc-core/src/trace.rs:31]: Don't we have an into for that ? if not plz implement the from trait

File: crates/client/rpc-core/src/trace.rs [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+/// Converts ContractAddress used in Blockifier/StarknetAPI to FieldElement from starknet-rs
+pub fn convert_contract_address(contract_address: ContractAddress) -> FieldElement {
+    convert_felt(*contract_address.0.key())
+}
+
```
Comment (LucasLvy) [inline crates/client/rpc-core/src/trace.rs:37]: same here

File: crates/client/rpc-core/src/trace.rs [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+/// Converts ClassHash used in Blockifier/StarknetAPI to FieldElement from starknet-rs
+pub fn convert_class_hash(class_hash: ClassHash) -> FieldElement {
```
Comment (LucasLvy) [inline crates/client/rpc-core/src/trace.rs:39]: same

File: crates/client/rpc-core/src/trace.rs [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    convert_felt(class_hash.0)
+}
+
+/// Converts CallType used in Blockifier/StarknetAPI to a local type compliant with the JSON-RPC
+/// spec
+pub fn convert_call_type(call_type: BlockifierCallType) -> CallType {
+    match call_type {
+        BlockifierCallType::Call => CallType::Call,
+        BlockifierCallType::Delegate => CallType::LibraryCall,
+    }
```
Comment (LucasLvy) [inline crates/client/rpc-core/src/trace.rs:49]: same

**Review by LucasLvy — CHANGES_REQUESTED** [review]

File: crates/client/rpc-core/src/trace.rs (hunk 5 above)
Comment (tdelabro) [reply crates/client/rpc-core/src/trace.rs:31]: All this will probably be change with my big type PR.
Wait for me to merge it and then do this job. But it won't be on the same types so, don't bother with if for now

File: crates/client/rpc-core/src/lib.rs (hunk 3 above)
Comment (m-kus) [reply crates/client/rpc-core/src/lib.rs:152]: You mean spec version or entire docstring?

File: crates/client/rpc-core/src/lib.rs [hunk 9]
```diff
@@ -145,4 +146,25 @@ pub trait StarknetRpcApi {
...
+    #[method(name = "simulateTransactions")]
+    fn simulate_transactions(
+        &self,
+        block_id: BlockId,
+        transactions: Vec<BroadcastedTransaction>,
+        simulation_flags: Vec<SimulateTransactionFlag>,
+    ) -> RpcResult<Vec<SimulateTransactionResult>>;
+
+    /// Compatibility with Starknet.js
+    /// See https://github.com/0xs34n/starknet.js/blob/829f5f4375a63c7c269e34dad22c9cd8874a64ba/src/provider/rpc.ts#L546
```
[the remark this answers is not in the document]
Comment (m-kus) [reply crates/client/rpc-core/src/lib.rs:162]: I think it was originally Pathfinder bug that propagated to starknet.js and now here as well :)

In `0.3.0` it was `simulateTransaction` with `transaction` field
In `0.4.0` it is `simulateTransactions` with `transactions` field
In starknet.js it is `simulateTransaction` with `transactions` field

File: crates/client/rpc-core/src/lib.rs (hunk 4 above)
Comment (m-kus) [reply crates/client/rpc-core/src/lib.rs:163]: In 0.3.0 it was `simulateTransaction` with `transaction` field
In 0.4.0 it is `simulateTransactions` with `transactions` field
In starknet.js it is `simulateTransaction` with `transactions` field

File: crates/client/rpc-core/src/trace.rs (hunk 5 above)
Comment (m-kus) [reply crates/client/rpc-core/src/trace.rs:31]: So what's the resolution, keeping these casts as is and remove later or implement `From` in starknet-api / blockifier (that would introduce additional dependency to those crates)?

**Comment by m-kus** [comment]
I will add few more tests to the pallet (to increase codecov and check some edge cases) also need to resolve failing integration tests.

File: crates/client/rpc-core/src/trace.rs (hunk 5 above)
Comment (tdelabro) [reply crates/client/rpc-core/src/trace.rs:31]: The best case is my PR is merged at the end of the week and you rebase on it next week.
If you don'y work on it full time and start with the other comments +  those on the other related PRs, it should be good. I hope
